### PR TITLE
General improvements for `module_programming_llm`

### PR DESCRIPTION
**Pull request by FelixTJDietrich**
### Motivation and Context

The `module_programming_llm` has been neglected a bit and I need to make it ready for presentation. This PR aims at making everything ready.

### Description

- Huge refactor of the module
- Use the model config from `module_text_llm` (just copy paste), also copied some other helpers 1 to 1
    * `module_programming_llm/helpers/llm_utils.py` copied
    * `module_programming_llm/helpers/models/*` copied
    * Env files are also copied
- Support function calling for OpenAI chat models and use structured output parsing similar to `module_text_llm`
- Remove programming language file extension limitation (LLMs can handle most file types anyways)
- Rework splitting of problem statement and grading instructions (it is now in the feedback generation pipeline, instead with receive submissions)
- General improvements to improve reliability and feedback quality

### Steps for Testing

1. Use `.env` file from `module_text_llm` in `module_programming_llm`
2. Start module and playground as always
3. Check out the new module's config if everything looks good in the playground
4. Generate feedback suggestions using `module_programming_llm` successfully (you don't have to test the Replicate models)

### CODE REVIEW

File: module_programming_llm/module_programming_llm/generate_suggestions_by_file.py [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+        for file_path in changed_files_from_template_to_submission
+    ]
+
+    # Changed text files
+    changed_files = load_files_from_repo(
+        submission_repo, 
+        file_filter=lambda file_path: file_path in changed_files_from_template_to_submission
+    )
+
+    for file_path, file_content in changed_files.items():
```
Comment (pal03377) [inline module_programming_llm/module_programming_llm/generate_suggestions_by_file.py:104]: A comment above this loop would be nice

File: module_programming_llm/module_programming_llm/generate_suggestions_by_file.py [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+            "template_to_submission_diff": template_to_submission_diff,
+            "template_to_solution_diff": template_to_solution_diff,
+            "grading_instructions": grading_instructions,
+            "problem_statement": problem_statement,
+        })
+    
+    # Filter long prompts (omitting features if necessary)
+    # "submission_file" is not omittable, because it is the main input containing the line numbers
+    # In the future we might be able to include the line numbers in the diff, but for now we need to keep it
+    omittable_features = [
```
Comment (pal03377) [inline module_programming_llm/module_programming_llm/generate_suggestions_by_file.py:157]: Maybe add a note that this list defines the order in which features are left out (priority). (If my understanding is correct)

File: module_programming_llm/module_programming_llm/helpers/llm_utils.py [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+    Returns:
+        Optional[T]: Parsed output, or None if it could not be parsed
+    """
+    if supports_function_calling(model):
+        chain = create_structured_output_chain(pydantic_object, llm=model, prompt=chat_prompt)
+        
+        try:
+            return await chain.arun(**prompt_input)
+        except (OutputParserException, ValidationError):
+            # In the future, we should probably have some recovery mechanism here (i.e. fix the output with another prompt)
```
Comment (pal03377) [inline module_programming_llm/module_programming_llm/helpers/llm_utils.py:135]: Add an error log here?

File: module_programming_llm/module_programming_llm/helpers/llm_utils.py [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+            return None
+
+    output_parser = PydanticOutputParser(pydantic_object=pydantic_object)
+    chain = LLMChain(llm=model, prompt=chat_prompt, output_parser=output_parser)
+    try:
+        return await chain.arun(**prompt_input)
+    except (OutputParserException, ValidationError):
+        # In the future, we should probably have some recovery mechanism here (i.e. fix the output with another prompt)
```
Comment (pal03377) [inline module_programming_llm/module_programming_llm/helpers/llm_utils.py:143]: Add an error log here?

File: module_programming_llm/module_programming_llm/helpers/models/replicate.py [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+from athena.logger import logger
+from .model_config import ModelConfig
+
+
+# Hardcoded list of models
+# If necessary, add more models from replicate here, the config below might need adjustments depending on the available
+# parameters of the model
+#
+# To update the version of the models, go to the respective page on replicate.com and copy the (latest) version id 
+# from and paste it after the colon in the value of the dictionary. Ever so often a new version is released.
```
Comment (pal03377) [inline module_programming_llm/module_programming_llm/helpers/models/replicate.py:17]: Good hint for future devs!

Comment (pal03377) [reply module_programming_llm/module_programming_llm/helpers/models/replicate.py:17]: We should also add this to the docs in the future

**Review by pal03377 — APPROVED** [review]
Code lgtm. I did not test it, but considering that a lot of stuff builds on this PR and that I will test the follow-ups again anyways I think it's ok if we merge this one 👍

File: module_programming_llm/module_programming_llm/helpers/models/replicate.py (hunk 5 above)
Comment (FelixTJDietrich) [reply module_programming_llm/module_programming_llm/helpers/models/replicate.py:17]: After everything is merged I will go through all my code and add stuff like this to the documentation. For example, how to add additional LLM providers etc.

File: module_programming_llm/module_programming_llm/helpers/llm_utils.py (hunk 4 above)
Comment (FelixTJDietrich) [reply module_programming_llm/module_programming_llm/helpers/llm_utils.py:143]: Implicitly logged to LangSmith (or our future logging solution) they also contain the errors raised. I will add the fallback/improved parsing to the future work items soon.

File: module_programming_llm/module_programming_llm/generate_suggestions_by_file.py (hunk 2 above)
Comment (FelixTJDietrich) [reply module_programming_llm/module_programming_llm/generate_suggestions_by_file.py:157]: I will add a note

**Pull request merged**